# 01 – Data Quality Assessment



This notebook performs a comprehensive data quality audit of the NovaCred credit application dataset. As the Data Governance Task Force, our objective is to systematically identify, quantify, and remediate issues that may compromise the integrity, reliability, and fairness of downstream machine learning decisions.



### DATASET OVERVIEW

In [109]:
# We import Libraries 
import pandas as pd 
import os
import numpy as np
import json 

In [113]:
with open('../data/raw_credit_applications.json', 'r') as f:
    data = json.load(f) 

print(f"Total raw records: {len(data)}")
df = pd.DataFrame(data)


Total raw records: 502


In [114]:
# We can use json_normalize to flatten the nested JSON structure into a DataFrame
df = pd.json_normalize(data)
df.head(100)

,_id,spending_behavior,processing_timestamp,applicant_info.full_name,applicant_info.email,applicant_info.ssn,applicant_info.ip_address,applicant_info.gender,applicant_info.date_of_birth,applicant_info.zip_code,...,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.loan_approved,decision.rejection_reason,loan_purpose,decision.interest_rate,decision.approved_amount,financials.annual_salary,notes
0,app_200,"[{'category': 'Shopping', 'amount': 480}, {'ca...",2024-01-15T00:00:00Z,Jerry Smith,jerry.smith17@hotmail.com,596-64-4340,192.168.48.155,Male,2001-03-09,10036,...,23,0.20,31212,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
1,app_037,"[{'category': 'Rent', 'amount': 608}, {'catego...",NaN,Brandon Walker,brandon.walker2@yahoo.com,425-69-4784,10.1.102.112,M,1992-03-31,10032,...,51,0.18,17915,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
2,app_215,"[{'category': 'Rent', 'amount': 109}]",NaN,Scott Moore,scott.moore94@mail.com,370-78-5178,10.240.193.250,Male,1989-10-24,10075,...,41,0.21,37909,True,NaN,vacation,3.7,59000.0,NaN,NaN
3,app_024,"[{'category': 'Fitness', 'amount': 575}]",NaN,Thomas Lee,thomas.lee6@protonmail.com,194-35-1833,192.168.175.67,Male,1983-04-25,10077,...,70,0.35,0,True,NaN,NaN,4.3,34000.0,NaN,NaN
4,app_184,"[{'category': 'Entertainment', 'amount': 463}]",2024-01-15T00:00:00Z,Brian Rodriguez,brian.rodriguez86@aol.com,480-41-2475,172.29.125.105,M,1999-05-21,10080,...,14,0.23,31763,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,app_323,"[{'category': 'Travel', 'amount': 133}, {'cate...",NaN,Michael Thompson,michael.thompson45@gmail.com,862-33-2796,172.26.169.208,Male,08/11/1981,10090,...,92,0.40,27746,True,NaN,NaN,3.9,80000.0,NaN,NaN
96,app_288,"[{'category': 'Shopping', 'amount': 434}]",NaN,Susan Young,susan.young22@mail.com,159-19-1765,192.168.114.250,Female,1962-11-05,90252,...,71,0.27,23054,True,NaN,NaN,5.5,67000.0,NaN,NaN
97,app_396,"[{'category': 'Shopping', 'amount': 698}, {'ca...",NaN,Katherine Harris,katherine.harris16@outlook.com,174-24-9597,10.14.75.95,Female,1976-01-28,10027,...,124,0.20,33232,True,NaN,NaN,5.4,40000.0,NaN,NaN
98,app_371,"[{'category': 'Rent', 'amount': 180}, {'catego...",NaN,Sarah Walker,sarah.walker95@mail.com,142-97-2132,10.57.113.229,Female,1995-02-23,90250,...,12,0.11,59032,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN


In [ ]:
# We investigate summary statistics for numerical fields
df.describe()

,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.interest_rate,decision.approved_amount,financials.annual_salary
count,500.000000,500.000000,500.000000,292.000000,292.000000,5.000000
mean,50.444000,0.245520,29579.530000,4.564726,47845.890411,69200.000000
std,31.289822,0.136148,16745.805308,1.162866,18103.754530,22664.950915
min,-10.000000,0.050000,-5000.000000,2.500000,15000.000000,45000.000000
25%,27.000000,0.150000,17387.250000,3.500000,34000.000000,46000.000000
50%,48.500000,0.230000,27399.000000,4.550000,48000.000000,75000.000000
75%,72.000000,0.342500,38281.750000,5.600000,62250.000000,86000.000000
max,133.000000,1.850000,88078.000000,6.500000,80000.000000,94000.000000


In [ ]:
# and a structured overview of the dataset
df.info()

<class 'pandas.DataFrame'>
Index: 500 entries, 383 to 152
Data columns (total 22 columns):
 #   Column                            Non-Null Count  Dtype              
---  ------                            --------------  -----              
 0   _id                               500 non-null    str                
 1   spending_behavior                 500 non-null    object             
 2   processing_timestamp              62 non-null     datetime64[us, UTC]
 3   applicant_info.full_name          500 non-null    str                
 4   applicant_info.email              500 non-null    str                
 5   applicant_info.ssn                496 non-null    str                
 6   applicant_info.ip_address         496 non-null    str                
 7   applicant_info.gender             500 non-null    str                
 8   applicant_info.date_of_birth      500 non-null    str                
 9   applicant_info.zip_code           500 non-null    str                
 10  fina

In [122]:
df.columns.tolist()

['_id',
 'spending_behavior',
 'processing_timestamp',
 'applicant_info.full_name',
 'applicant_info.email',
 'applicant_info.ssn',
 'applicant_info.ip_address',
 'applicant_info.gender',
 'applicant_info.date_of_birth',
 'applicant_info.zip_code',
 'financials.annual_income',
 'financials.credit_history_months',
 'financials.debt_to_income',
 'financials.savings_balance',
 'decision.loan_approved',
 'decision.rejection_reason',
 'loan_purpose',
 'decision.interest_rate',
 'decision.approved_amount',
 'financials.annual_salary',
 'notes']

In [ ]:
# We analyze the decision outcome distribution
approved = (df["decision.loan_approved"] == True).sum()
rejected = (df["decision.loan_approved"] == False).sum()
total = len(df)

print(f"Approved: {approved} ({approved/total*100:.1f}%)")
print(f"Rejected: {rejected} ({rejected/total*100:.1f}%)")
print(f"Total:    {total}")

Approved: 292 (58.2%)
Rejected: 210 (41.8%)
Total:    502


### Overview Findings 

The dataset containts 502 credit application records that include applicant information, financial attributes, spending behavior categories and loan decision outcomes. 


The decision outcome is balanced - with 58.2% approved applications and 41.8% rejected. 

However some statitistics caught our eye:
- annual_income --> stored as object type rather than numeric - meaning that some string values are affecting data consistency 
- credit_history_months --> the minimum value yielded -10 - implausible for credit history values 
- savings_balance --> has a minimum of -5,000 - not common for a balance field 

### MISSINGNESS ANALYSIS

In [ ]:
# We check for missing values in each column + percentage of missing values per column
NaN_investigation = pd.DataFrame({'Missing_Count': df.isnull().sum(), 
                                  'Missing_Percent': (df.isnull().sum() / len(df) * 100).round(2)
                                  }).sort_values(by='Missing_Count', ascending=False)

display(NaN_investigation)

,Missing_Count,Missing_Percent
notes,500,99.60
financials.annual_salary,497,99.00
loan_purpose,452,90.04
processing_timestamp,440,87.65
decision.rejection_reason,292,58.17
decision.approved_amount,210,41.83
decision.interest_rate,210,41.83
financials.annual_income,5,1.00
applicant_info.ip_address,5,1.00
applicant_info.ssn,5,1.00


### Missingness Findings

Many columns showed high missingness rates. Outside the spending variables, the columns: Notes, Annual_salary, Loan_purpose and Processing_timestamp have the highest rates.
We will address this with focused analysis.

## Data Quality Issue 1 - Duplicate Records


In [129]:
# We investigate the duplicate records by ID (the primary key that needs to be unique for each application)
duplicate_count = df["_id"].duplicated().sum()
print(f"Duplicate ID count: {duplicate_count}")
display(df[df["_id"].duplicated(keep=False)].sort_values("_id"))

Duplicate ID count: 2


,_id,spending_behavior,processing_timestamp,applicant_info.full_name,applicant_info.email,applicant_info.ssn,applicant_info.ip_address,applicant_info.gender,applicant_info.date_of_birth,applicant_info.zip_code,...,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.loan_approved,decision.rejection_reason,loan_purpose,decision.interest_rate,decision.approved_amount,financials.annual_salary,notes
383,app_001,"[{'category': 'Fitness', 'amount': 576}]",NaN,Stephanie Nguyen,stephanie.nguyen47@mail.com,427-90-1892,10.121.120.213,Female,1986-05-27,90230,...,37,0.42,0,False,high_dti_ratio,NaN,NaN,NaN,NaN,NaN
455,app_001,"[{'category': 'Fitness', 'amount': 576}]",NaN,Stephanie Nguyen,stephanie.nguyen47@mail.com,NaN,NaN,NaN,NaN,NaN,...,37,0.42,0,False,high_dti_ratio,NaN,NaN,NaN,NaN,DUPLICATE_ENTRY_ERROR
8,app_042,"[{'category': 'Insurance', 'amount': 153}, {'c...",NaN,Joseph Lopez,joseph.lopez1@gmail.com,652-70-5530,192.168.91.142,Male,1990-05-04,10044,...,43,0.41,15974,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
354,app_042,"[{'category': 'Insurance', 'amount': 153}, {'c...",NaN,Joseph Lopez,joseph.lopez1@gmail.com,652-70-5530,192.168.91.142,Male,1990-05-04,10044,...,43,0.41,15974,False,algorithm_risk_score,NaN,NaN,NaN,NaN,RESUBMISSION


Two applications IDs are duplicated, in both cases the records are mostly identical with some exceptions.
This is a critical issue because the "_id" field is supposed to be a unique identifier for each record to ensure traceability and auditability.

In [ ]:
# we found 2 duplicate records based on the "_id" column. 
# duplicates violate uniqueness constraints and compromise data integrity

duplicates = df[df["_id"].duplicated(keep=False)].sort_values("_id")

# we can see that some records have more missing values than others. We will keep the most complete record for each duplicate ID.
df["completeness_score"] = df.notnull().sum(axis=1)


df = df.sort_values(
    by=["_id", "completeness_score"],
    ascending=[True, False]
)

# we remove both the duplicate records and the completeness score column
df = df.drop_duplicates(subset="_id", keep="first")
df = df.drop(columns=["completeness_score"])

print("Remaining duplicate IDs:", df["_id"].duplicated().sum())

Remaining duplicate IDs: 0


## Data Quality Issue 2 – Column Inconsistency (annual_income vs annual_salary) 


The 5 records missing annual_income are exactly the 5 records where annual_salary is populated instead. No record has both fields, and no record is missing income data entirely. 

This is a schema inconsistency: the same financial attribute was recorded under two different field names, likely due to a change in the data collection pipeline or inconsistent field mapping across intake systems.

In [90]:
df[df["financials.annual_salary"].notna()][
    ["financials.annual_income", "financials.annual_salary"]
]

# schema duplication or incomplete data pipeline integration - we fill income with salary only where income is missing (5 records
df["financials.income_unified"] = (
    df["financials.annual_income"]
    .combine_first(df["financials.annual_salary"])
)

# we check the number of records where income_unified is still missing
missing_income_unified = df["financials.income_unified"].isnull().sum()
print(f"Records with missing income_unified: {missing_income_unified}")



Records with missing income_unified: 0


## Data Quality Issue 3 – Categorical Encoding


In [91]:
# Analyze categorical encoding consistency for gender
gender_counts = (
    df["applicant_info.gender"]
    .fillna("NaN")
    .replace("", "Empty String")
    .value_counts(dropna=False)
    .reset_index()
)

gender_counts.columns = ["Gender Value", "Count"]

display(gender_counts)


,Gender Value,Count
0,Male,194
1,Female,193
2,F,58
3,M,53
4,Empty String,2


Categorical fields should have consistent encoding across all records. Variations complicate both aggregation and analysis - the goal is to standardize such inconsistencies

In [92]:
# we standardize the gender column to ensure consistent values M = male  
def standardize_gender(df):
    df["applicant_info.gender"] = (
        df["applicant_info.gender"]
        .astype(str)
        .str.strip()
        .str.lower()
    )
   
    df["applicant_info.gender"] = df["applicant_info.gender"].replace({
        "m": "male",
        "male ": "male",
        "f": "female",
        "female ": "female"
    })
    
    df.loc[df["applicant_info.gender"] == "nan", "applicant_info.gender"] = np.nan
    
    return df

df = standardize_gender(df)

# we check the value counts for gender to verify the standardization
df["applicant_info.gender"].value_counts()

applicant_info.gender
female    251
male      247
            2
Name: count, dtype: int64

## Data Quality Issue 4 – Data Inconsistencies in Time Columns


In [ ]:
# We check the unique values in the time columns to identify inconsistencies

# 440/500 records (88%) missing processing_timestamp
# we standardize the time format in processing_timestamp column to another format
df["processing_timestamp"] = pd.to_datetime(
    df["processing_timestamp"],
    errors="coerce",
    utc=True
)

print("Unique processing_timestamp values after standardization:")
print(df["processing_timestamp"].unique())

# we check for future timestamps in processing_timestamp column, which would indicate data entry errors
(df["processing_timestamp"] > pd.Timestamp.today(tz="UTC")).sum()

Unique processing_timestamp values after standardization:
<DatetimeArray>
[                      'NaT', '2027-01-20 00:00:00+00:00',
 '2024-01-15 00:00:00+00:00', '2026-03-15 00:00:00+00:00',
 '2025-12-01 00:00:00+00:00']
Length: 5, dtype: datetime64[us, UTC]


np.int64(2)

Inconsistent date formats complicate parsing, comparison, and temporal analysis - standardization is required to avoid data ambiguity. 

## Conclusions 

The quality assessment identified violations of business rules in the dataset. 



The most critical findings are:

1. Three loans were approved without documented income.Income is a core underwriting variable; its absence at the point of decision indicates a control gap in the approval workflow that cannot be remediated retrospectively.

2. Two application IDs are duplicated, affecting 4 records. Primary key violations undermine traceability, audit reliability, and the integrity of any downstream joins or aggregations.

3. Empty strings in identity and demographic fields are not detected as missing by standard validation checks, causing true incompleteness to be systematically undercounted across email, gender, date_of_birth, and zip_code.

4. 87.6% of records lack a processing timestamp, severely weakening the audit trail for the credit decision lifecycle.

## Export Cleaned Dataset

In [94]:
# Export cleaned dataset for Bias and Governance analysis

# Final column 
df["financials.income_unified"] = (
    df["financials.annual_income"]
    .combine_first(df["financials.annual_salary"])
)

# Drop useless columns 
columns_to_drop = [
    "financials.annual_salary",
    "financials.annual_income",
    "notes"   # we drop notes columns since it has mostly missing values 
]

df_clean = df.drop(columns=[col for col in columns_to_drop if col in df.columns])

if "processing_timestamp" in df_clean.columns:
    df_clean["processing_timestamp"] = pd.to_datetime(
        df_clean["processing_timestamp"],
        errors="coerce",
        utc=True
    ).dt.strftime("%Y-%m-%dT%H:%M:%SZ")

df_clean.to_csv("cleaned_dataset.csv", index=False)